### DISCUSSION TITLE
Three. Binding Model and Implementation

### DISCUSSION THREAD

**Discussion by Mehdi-H** [Mehdi-H/DDD-blue-book-collaborative-reading#4] [Chapter]
# Three. Binding Model and Implementation

You can discuss here about this chapter ! 👇

Ask questions,
or share anything you find insightful !

**Comment by sofcalca** [comment]
## :construction: Glossary
MODEL-DRIVEN DESIGN

## What I learned in this chapter

### Analysis model

It is a classical way of using the model:
- useful at the beginning of a project to discover the domain
- but it is not updated: the business and its code implementation can be misaligned
- what is learned by the developers is not integrated to the model: so non-techs might not get the information
- it is misleading: non-techs might thing that the software does something it does not because of these divergence

### Model-driven design

DDD sees the model beyond its analysis capabilities. It now has two goals:
- be a tool for analysis
- be a foundation for the design

> Development becomes an iterative process of refining the model, the design, and the code as a single activity

To achieve this:
- the code follows the model and transforms it : 
    - it is created aligned with the business goal
    - it is updated when the implementation discovers something
- object-oriented programming makes this simpler
    - object design: the code should express the concepts of the model

Consequence:
- all techs that participate in the modeling process should participate in its code implementation
- all developers should be in contact with domain experts to contribute to the model
- every person, tech and non-tech, should discuss with techs using ubiquitous language

_All quotes are from Eric Evan's book Domain Driven Design_

**Reply by Mehdi-H** [reply]
> it is misleading: non-techs might thing that the software does something it does not because of these divergence

I really like the given example of "Microsoft Internet Explorer Favorites" in this chapter, it's a good example of **leaky abstraction**[^1], and this chapter helped me identify that **focusing only on analysis model and set model-driven design activities aside can be one root cause of leaky abstraction**.

I note that consequences of doing _only_ analysis modeling can also be misleading for a technical population

[^1]: Spolsky, Joel (2002). ["The Law of Leaky Abstractions"](https://www.joelonsoftware.com/2002/11/11/the-law-of-leaky-abstractions/)
